### ISSUE TITLE
Consider showing properties on the file list

### ISSUE THREAD

**Issue by rmcanany** [rmcanany/SolidEdgeHousekeeper#139]
Suggested by @Careframe1 on the SE Forum:
https://community.sw.siemens.com/s/question/0D5Vb00000Lej9IKAR/solid-edge-housekeeper-v20243-is-released

Also be able to sort on property values.

**Comment by farfilli** [comment]
I considered doing that some time ago, but then I got busy and didn't try it.

For sure don't even try to use the Solid Edge property APIs, it would have a bad performance impact on a large list of files.
I suggest using the same method as in the direct properties editor; it should be fast enough.

I can take this task.

**Comment by rmcanany** [comment]
OK, I assigned it to you.  No rush, obviously.

**Comment by rmcanany** [comment]
I'm thinking about changing ListOfColumns to store the whole property formula (eg %{System.Title}) instead of just the name (Title).  

Two reasons:
1.  We could reuse UtilsCommon.GetOLEPropValue(), eliminating UtilsFileList.FindProp.
2.  We could provide the option to use DM if someone wanted Material or other properties not in structured storage, and "|R1" props for Draft files.

The second option would incur a performance hit, but the user could decide if it's worth it.

I can make the change.  I'm mostly wondering if you see anything I'm missing.  (BTW, I would strip the prop name from the formula for the column header text.)

**Comment by farfilli** [comment]
Yes Robert sounds good. I would store it in a duupe so we can also store the visibility.
It would be a list(of String, Boolean)

**Comment by rmcanany** [comment]
Cool.  I'll play around with it today.

**Comment by farfilli** [comment]
Partially completed with #143 

A possible future improvement could be to store column width and ability to reorder columns

**Comment by farfilli** [comment]
@rmcanany On the Column Properties localized ones don't work, I have to add manually the English name. Can you take a look?

**Comment by rmcanany** [comment]
@farfilli I will take care of that.  

I am going to use UtilsCommon.SubstitutePropertyFormula, which already does it.  I'm also going to add a Formula property to the PropertyColumn class.  It would be in the correct format, eg %{System.Title}.

I like the magic delete buttons, by the way.  That slick how you did that.

On a related note, what does this do?
`        Me.CLB_Properties.Items.AddRange(New Object() {"Name", "Path"})`

**Comment by farfilli** [comment]
It had the initial two properties name and path to the checkboxlist

**Comment by rmcanany** [comment]
Oh, got it.

**Comment by farfilli** [comment]
But they start unchecked, perhaps add them programmatically at first startup or if they miss in the configuration file. It needs to be initialized in some way.

About the magic button its just a regular button that show/hide based on mouse position

**Comment by rmcanany** [comment]
OK, I'll make sure they're initialized correctly.  That's no problem.

**Comment by rmcanany** [comment]
@farfilli 
I added Formula to PropertyColumn and switched from using FindProp to SubstitutePropertyFormula.  When you get a chance, could you test localized properties?

I haven't done the saved settings yet.  That's next.

**Comment by farfilli** [comment]
Localized properties still doesn't work, in the image you can see the localized one beside the english one

![image](https://github.com/user-attachments/assets/b00a8a81-dd98-4c60-bfe5-9e814802663b)

**Comment by rmcanany** [comment]
I added an initializer for template properties.  It's up on GitHub.  Not sure that's the whole story.  Let me know if it works.  

While you're at it, it would be good to check Edit Properties, too.  It uses the same code and might tell me if something is broken in there.

**Comment by farfilli** [comment]
The ListOfColumns List doesn't contain "Name" and "Path" anymore; they are mandatory to be in the list at positions 0 and 1.
It seems they are missing now therefore it is not possible to show/hide them.
The column configuration is not maintained from one run to another

Localization still not working
![image](https://github.com/user-attachments/assets/ad3f4b0c-a313-47cb-acba-888a3e779e2e)

**Comment by farfilli** [comment]
Fixed Path and Name startup with #144

**Comment by rmcanany** [comment]
Thanks for that.  I'm still working on the saved settings part.

**Comment by rmcanany** [comment]
Can do.  To save me some digging, do you know off the top of your head how to get the widths?

**Comment by rmcanany** [comment]
Oh sorry, I meant get/set on the `ListView` object.  I found it: `ListViewFiles.Columns.Item(i).Width = W`

**Comment by rmcanany** [comment]
I think the saved settings are working now.  I haven't added the column width piece yet.

**Comment by farfilli** [comment]
It seems to work, good job

**Comment by farfilli** [comment]
I have implemented the column width property with #145 

I did try to implement the width saving but as I understand if I update the JSON list then an update is fired that goes into a loop.
I blocked the loop initially by a boolean property, then a better method was to compare width values and if they don't really change skip the update.

**Comment by farfilli** [comment]
Yes I'm done but don't forget that localized properties still doesn't work

**Comment by rmcanany** [comment]
Oh, I missed the part about localized properties.  Did you say they're working on Edit Properties?

**Comment by rmcanany** [comment]
Never mind.  I found the issue.  Now to fix it...

**Comment by rmcanany** [comment]
I don't know if it's working, but at least it is compiling.  

If it works in the list properties code, it should work everywhere.  If you want to double-check, here are the other places SubstitutePropertyFormula is called.
- TaskEditProperties: FindText and ReplaceText
- SaveDrawingAs: SubdirectoryFormula
- SaveModelAs: SubdirectoryFormula

**Comment by rmcanany** [comment]
What do you think about Export List including the extra columns?

**Comment by farfilli** [comment]
> I don't know if it's working, but at least it is compiling.
> 
> If it works in the list properties code, it should work everywhere. If you want to double-check, here are the other places SubstitutePropertyFormula is called.
> 
> * TaskEditProperties: FindText and ReplaceText
> * SaveDrawingAs: SubdirectoryFormula
> * SaveModelAs: SubdirectoryFormula
> 
> **Edit**: The SaveAs commands don't use structured storage, so they should be working already. Hopefully.

I have tested localized properties in Direct Edit and Solid Edge Edit on FindText, and ReplaceText both in PT and in EX and it seems to work fine.

Tested SaveModelAs in SubdirectoryFormula and it works ok

There is a bug with Custom properties that seems to be case-sensitive:

These settings work fine:
![image](https://github.com/user-attachments/assets/2bdedc5f-8350-48fd-8333-09ced40b1e06)

These settings lead to an error:
![image](https://github.com/user-attachments/assets/5838152a-7cad-49c7-af72-846069beb3c1)
![image](https://github.com/user-attachments/assets/37f85e96-aab2-4ebd-9f72-232f156dec15)

Please note that in the file the custom properties is written in upper-case "OGGETTO 1"

**Comment by farfilli** [comment]
I did try to implement the column update after the property edit but in the case of Solid Edge edit I got a cross-thread error; not a problem in direct edit. I did try to pass the listviewitem as a parameter in the Process sub but it wasn't a good idea.

I think the best approach would be the thread passing back info to update the listviewitem values after each execution, it may be something to insert in the dictionary of errors, which would be a dictionary of execution or something like that.

**Comment by farfilli** [comment]
@rmcanany I have implemented property edit from FileListView, just double-click a property to edit it.
I have created a new method in TaskEditProperties to edit a single file property

**Comment by rmcanany** [comment]
Good deal.  I like the idea of editing the properties like that, but I got an exception.  The file isn't open in another application as far as I know.

![exception](https://github.com/user-attachments/assets/1cc6a38f-2438-44eb-8ffb-b4394e50851e)


Not sure how to handle the text case issue.  It's bombing out because `TemplatePropertyDict` does not have that key.  I guess I could convert everything to upper case.  I would have to see what other problems that might create.  

Is something like that going on in `UC.GetOLEPropValue()`?  Could you explain what `Function(Proper) Proper.PropertyName` does?  Also why sometimes upper case and other times lower case?

`OLEProp = System_Properties.Properties.First(Function(Proper) Proper.PropertyName = "PIDSI_" & PropertyName.ToUpper)`
`OLEProp = System_Properties.Properties.FirstOrDefault(Function(Proper) Proper.PropertyName.ToLower Like "*" & PropertyName.ToLower & "*")`
`OLEProp = Custom_Properties.UserDefinedProperties.Properties.FirstOrDefault(Function(Proper) Proper.PropertyName = PropertyName)`

I've run across the cross-thread issue before.  A programmer I know said that's what `STDOUT` and friends are for.  Once I get it working for error reporting, I can look at updating the ListView.

**Comment by farfilli** [comment]
I don't know what may lock the file, I did test with some files in a folder and was working correctly. Didn't test it extensively.

About the ToUpper is mandatory in streams cause they are written that way. The ToLower is just an easy way to compare it could be ToUpper as well.

About the custom properties being case sensitive I didn't investigate the reason,

**Comment by farfilli** [comment]
That Function is to get the property by name without the need to parse all the properties stream.

**Comment by rmcanany** [comment]
Got it.  

Changing a property in the ListView works for me sometimes and not others.  I'll see if I can narrow it down.

Is there a reason for `UpdateSingleProperty` to be in `TaskEditProperties`?  I'm thinking about moving it to  `UtilsCommon`.

**Comment by farfilli** [comment]
I didn't know how the localization worked and thought it had to be there. Move it where you think is it's place 👍🏻

**Comment by farfilli** [comment]
Need to implement the addition of custom properties in case they are missing

**Comment by rmcanany** [comment]
Forgot about that one.  I can play around with it.

On a related note, I downloaded SSView to get a better idea about structured storage.  Poking around, I noticed that Material, Material Thickness, and Sheet Metal Gage are all stored in there.  For some reason I thought they weren't.  Was the issue with OpenMcdf not being able to access the correct stream or something?

**Comment by farfilli** [comment]
I haven't found the density, that's the main concern about changing material.

All other things you found are already managed by direct editing.

I can do it on Monday, It was just a reminder 😅

**Comment by rmcanany** [comment]
Oh, I didn't know that direct edit had access to all those.  Great!  

Do we need to update these lists, or is it handled some other way?
```
        Dim SIList As New List(Of String)
        SIList.AddRange({"Title", "Subject", "Author", "Keywords", "Comments"})

        Dim DSIList As New List(Of String)
        DSIList.AddRange({"Category", "Company", "Manager"})

        Dim FunnyList As New List(Of String)
        FunnyList.AddRange({"Document Number", "Revision", "Project Name"})
```


I found Density with SSView:
![density_in_structured_storage](https://github.com/user-attachments/assets/4ba7f0ed-4755-4cb5-96d2-941e3ab4d45e)

I would be glad to let you handle the OpenMcdf stuff.  I reread your Issue on their GitHub page.  I could maybe figure it out, but not in this lifetime.

**Comment by farfilli** [comment]
I Will take a look at that, it opens to more capabilities 👍🏻

**Comment by rmcanany** [comment]
Sounds good.  

I'm still getting random exceptions when editing the ListView and now when doing Edit Properties outside of SE.  For Edit Properties it's random.  Run it two times in a row and different files bomb out.  It's usually 3 or 4 files out of a hundred.  Can you think of something I could have messed up on my end?

**Comment by farfilli** [comment]
Did you try with the file in the training folder?

**Comment by rmcanany** [comment]
Same story in the training folder.  That's a good idea about getting the latest build.  How do I do that?

**Comment by rmcanany** [comment]
I downloaded the zip files.  Do I just place the dlls in My Project folder?

**Comment by rmcanany** [comment]
Come to think of it, it's happening when the file stream is being created -- before OpenMcdf even gets its hands on it.  I found a bunch of info on the issue [here](https://stackoverflow.com/questions/26741191/ioexception-the-process-cannot-access-the-file-file-path-because-it-is-being).  I might try a few of those suggestions.

**Comment by rmcanany** [comment]
I got to work 2 times in a row by adding these lines in two locations in UtilsFileList.  It seems to have fixed ListView edit as well as Edit Properties.
```
cf.Close()
fs.Close()
cf = Nothing
fs = Nothing
```

**Comment by farfilli** [comment]
> I downloaded the zip files.  Do I just place the dlls in My Project folder?  Also, net40 or netstandard2.0?

I usually download the latest code, build the DLLs and copy them in the project folder (overwriting the current ones)

should be net40

**Comment by farfilli** [comment]
> Need to implement the addition of custom properties in case they are missing

I have implemented it by 8a6320c205f31162e8532c22e8a087ce801f9d61

**Please note that there is an important code change**

**Comment by farfilli** [comment]
> Oh, I didn't know that direct edit had access to all those. Great!
> 
> Do we need to update these lists, or is it handled some other way?
> 
> ```
>         Dim SIList As New List(Of String)
>         SIList.AddRange({"Title", "Subject", "Author", "Keywords", "Comments"})
> 
>         Dim DSIList As New List(Of String)
>         DSIList.AddRange({"Category", "Company", "Manager"})
> 
>         Dim FunnyList As New List(Of String)
>         FunnyList.AddRange({"Document Number", "Revision", "Project Name"})
> ```
> 
> I found Density with SSView: ![density_in_structured_storage](https://private-user-images.githubusercontent.com/64286960/375971662-4ba7f0ed-4755-4cb5-96d2-941e3ab4d45e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg4ODY4OTgsIm5iZiI6MTcyODg4NjU5OCwicGF0aCI6Ii82NDI4Njk2MC8zNzU5NzE2NjItNGJhN2YwZWQtNDc1NS00Y2I1LTk2ZDItOTQxZTNhYjRkNDVlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDE0VDA2MTYzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJlYTA0MDQ1MTQzZWUyMDVhNGRkNTRjOTZlZTIwMjczNTI4N2VkYmMwNDdhZTYxZThhMjZkMGYyOGFjZmI1NGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ToylyMP2bq2SRIhRojWcGRNVHU-lxWtAfTi-Pk-Pves)
> 
> I would be glad to let you handle the OpenMcdf stuff. I reread your Issue on their GitHub page. I could maybe figure it out, but not in this lifetime.

The density and Accuracy shown in the picture are present in the custom list as exposed but can be deleted.
I noticed that they are recreated if you don't have them while changing the material through the Solid Edge interface
If you change the value of those properties instead the Material is not updated, in my opinion, those properties should be read-only.

Changing the density through SSView doesn't work, I didn't try with OpenMCDF, in any case, it would not update the Material density.
Changing the name of the material will not update the other physical properties therefore a method to read the mtl file and set the properties would be required, and in the end, the Density will not be updated cause I haven't found it in the material section.
![image](https://github.com/user-attachments/assets/d2c64fe7-5a81-4562-bdb9-e22a888961b8)

We could instead implement perhaps those properties as read-only for reference purposes, let me know what do you think

**Comment by rmcanany** [comment]
OK.  That makes sense -- those are exposed variables, so probably don't propagate back to the variable table.  I like the read-only idea.

For the release I'm working on, do you think we're ready to include the changes to the ListView?  If not, I think I can just disable the button on the toolbar.  Let me know.

**Comment by rmcanany** [comment]
By the way, I checked one file where I had changed a custom property and it showed up correctly in the Windows property page.

**Comment by farfilli** [comment]
Did you change it before the last update of the DLL ?

**Comment by rmcanany** [comment]
Agreed.  I'm going to disable the button for now.

I didn't update the DLL on my end.  Did you update it on your side?

**Comment by farfilli** [comment]
I did update it on 09/09/2024, can you check if your edit occurred previously?

We need to create a list of checks to do when updating these DLLs cause they are evolving fast.
When the next release will be on NuGet we will move on it and consider it not experimental anymore.

**Comment by rmcanany** [comment]
My edit was in the last couple of days.  Maybe it has something to do with localized properties?

**Comment by farfilli** [comment]
> My edit was in the last couple of days. Maybe it has something to do with localized properties?

Custom properties shouldn't be affected by localization, does the code try with them too?

**Comment by rmcanany** [comment]
OK, so that's not in the released version, right?

**Comment by rmcanany** [comment]
Oh, right about custom properties.  For those it still uses the lookup table, but the English name is the same as the custom property name.

**Comment by rmcanany** [comment]
I got the click event to fire by changing `HeaderStyle` to Clickable.  

If you have time, you're welcome to do the sorting.  Otherwise I think I can do it with [this sample code](https://learn.microsoft.com/en-us/dotnet/api/system.windows.forms.listview.columnclick?view=windowsdesktop-8.0).

**Comment by rmcanany** [comment]
I kinda got the sort code started, but am having trouble with SubItems comparison.  Maybe there's an easier way.  Ideas?

**Comment by rmcanany** [comment]
And once you're done debugging that one, I have another.

I am trying to consolidate the fetching of an OLEProp.  For a test, I modified UpdateSingleProperty to use my new GetOLEProp consolidated code.  The idea was to eventually do the same in GetOLEPropValue, and TaskEditProperties.

In UpdateSingleProperty, if you set MyWay = False (in other words, do it your way) it works.  Set MyWay = True and it doesn't.  I must not be saving it correctly.  I used co.Save(dsiStream) and cf.Commit(), but no luck.  If you get a chance, could you see if you can spot where I'm going wrong?

The way I check if it worked, by the way, was to change a property value on ListViewFiles, then update the list.

**Comment by farfilli** [comment]
> I kinda got the sort code started, but am having trouble with SubItems comparison. Maybe there's an easier way. Ideas?

Done with #148 , I had that code ready yesterday sorry for the late commit

**Comment by rmcanany** [comment]
That looks great!  Nice job.  This is done, right?

I'll add a separate issue about cleaning up the DLLs.  Maybe another on consolidating OLEProps.

**Comment by rmcanany** [comment]
I've gotten this exception a couple of times.  I'm pretty sure it's only after I hit the Cancel button.

![image](https://github.com/user-attachments/assets/b3117f38-5c14-4104-8841-9ca1830144ac)

**Comment by farfilli** [comment]
It's the same issue I got on end form. Perhaps it's the same issue, it occur with collapsed ground?

**Comment by rmcanany** [comment]
Nothing was collapsed in the ListView.  Could I just add a Try/Catch there?

**Comment by rmcanany** [comment]
I am getting ready to make Material available as an OLEProp.  (I don't really think the others are necessary, but could easily add them if you do.)  

How do you map from a Property Set to a Stream?
![image](https://github.com/user-attachments/assets/2ab3b575-9c5a-4a72-b5c3-ea33982d805d)

**Comment by farfilli** [comment]
Have a look here:

https://github.com/ironfede/openmcdf/commit/0c8fc18a3046f0d7868303d5b279f0bff5344e66#comments

**Comment by farfilli** [comment]
> Nothing was collapsed in the ListView. Could I just add a Try/Catch there?

Today the problem didn't appear, have you solved it? I removed the column expand before close

**Comment by farfilli** [comment]
It would be good to save the MainForm size and position on close and restore on load

**Comment by farfilli** [comment]
I have implemented Extended Properties and the code to handle different property types.

Please have a look at my comments in the code and here https://github.com/rmcanany/SolidEdgeHousekeeper/pull/153#issuecomment-2416673514

**Comment by rmcanany** [comment]
Save and restore form size and position complete.

**Comment by rmcanany** [comment]
Added protections to keep a column from referring to an unknown property.  The combo box is now a non-editable drop down list.  I still need to add similar functionality to other places where properties can be used, e.g. Edit Properties, Save As, etc.

![image](https://github.com/user-attachments/assets/ca78b5f5-4a5f-4ade-928f-6b4dfdb9817d)

**Comment by rmcanany** [comment]
If I understood your suggestion to consolidate the different read-write property functions, there is already a method, `UC.TemplatePropertyDictAddProp` that does the writing part.  It handles the localization piece by consulting a lookup table.  

Is that the kind of thing you're looking for, or are you talking about something else?

**Comment by rmcanany** [comment]
Added `Property Selector` to `Property Filter` and `Edit Variables` (which now accepts a property in the Formula field).  Right-click the appropriate text box to open the dialog.

Added a Design Manager option to `UC.SubstitutePropertyFormula`.  Uses the new `UC.GetDMProp` to get those properties.

Added a new key to `TemplateProperyDict`, `PropertySetActualName`, to avoid the confusion between e.g. `SummaryInformation` and `System`.  The only place it is used is in the localized property lookup table.  There it's required.  Delete `Preferences\form_main_settings.json` to get around the exception that causes.

**Comment by farfilli** [comment]
I got some error while loading the old property columns configuration file at startup.
In my opinion, this situation should be handled so the user won't lost its configuration when moving from an HK version to a newer

**Comment by rmcanany** [comment]
I'll have to see if I can fix it.  Could it be related to the new dictionary key I mentioned?

**Comment by farfilli** [comment]
> I'll have to see if I can fix it. Could it be related to the new dictionary key I mentioned?

Yes, I think that's the problem.

Some Thoughts:
In my opinion instead of a dictionary I would construct a class to handle the whole columns thing, it would be easier to invoke methods like load\save\update and easy to expand in future functionality.
In general, I see you use a large amount of dictionaries that lead to a lot of code, it would be good to start creating specific classes for easy code management.

You wrote almost all the code so you know how to find things and how it works but for me, it is difficult to follow it in case I need to change or expand it. Perhaps one day we can plan a whole new release rewritten with all the features in mind. 😄

**Comment by farfilli** [comment]
Showing a lot of properties in the listview has an impact on the update button.

For example, for a list of 2560 files on my NAS, an update takes 0,9 minutes without properties, 2,4 minutes with 2 properties, and 5,1 minutes with 5 properties

I would add a warning somewhere somehow.

I have already optimized the openMCDF code to avoid opening and closing multiple times in the same file but I suspect there is space for some more improvement. If you have time and want to analyze where most of the time is spent you are welcome.

There has been some enhancement in open MCDF later in performance, I will test it out to see what score we get.

**Comment by rmcanany** [comment]
Appreciate the suggestions.  I like the idea of using classes.  It's easier to find how it's constructed, for one thing.  Also, you can mix Types which is something I have to work around now.  It's also also something that can be implemented over time.  I might start with TemplateProperties and see how it goes.

I'll add a warning in the Readme about performance.  Of course we want it to be fast, but the speed doesn't really bother me too much.  Showing progress on the Status bar lets me know it's doing *something*, which helps.

**Comment by farfilli** [comment]
> I'll add a warning in the Readme about performance. Of course we want it to be fast, but the speed doesn't really bother me too much. Showing progress on the Status bar lets me know it's doing _something_, which helps.

I often work with large list of files, like 40000, and wait an hour just to populate the list because I forgot property columns is not ideal. Maybe a warning could pop-up if you have a list over 1000 files that ask you if you want to disable the property columns.
To make it work, it would first retrieve the list of files and then eventually update the property columns.

**Comment by farfilli** [comment]
One more note, I have implemented backcolor in the properties.
White color = regular property 
Gray color = property not present
Pink color = can't read it, the file is open by someone

**Comment by rmcanany** [comment]
I saw that color coding.  Nice.  Did you see my note in your latest PR about the properties not showing up for me?

Makes sense about the large list of files.

**Comment by farfilli** [comment]
> Did you see my note in your latest PR about the properties not showing up for me?

I missed that one, where is it?

**Comment by rmcanany** [comment]
Here in the comments: https://github.com/rmcanany/SolidEdgeHousekeeper/pull/154

**Comment by farfilli** [comment]
Check during the properties retrieval what values it looking for.

Can you also send me a couple of those files?

In the end, I'm thinking to populate properties on listview item visibility to gain much in performance

**Comment by rmcanany** [comment]
I added a check for a large number of files and provided a message box for the user to decide if properties should be included or not.

I also changed around how FoundFiles are sent to the list view update.  Now files from all Sources are gathered into one list.  Duplicates are removed before updating the list, eliminating the time-consuming check for duplicates there.

On a test with 29,692 files and no processing of properties, the list updated in 0.2 min.  When processing with 2 properties, it took 7.6 min.

I like the idea of skipping updates for properties that are not displayed.  If a user decides to display a new column, we could set the ListViewFilesOutOfDate flag.  That way they couldn't proceed without updating the list.

**Comment by rmcanany** [comment]
I will check what's happening when trying to retrieve those properties on my files.

**Comment by rmcanany** [comment]
Just did a quick test and they're showing up now.  I may have been looking a draft files before.  

Actually, one file didn't show the properties.  I checked and they're present in the file.  I'll see if I can figure out what's going wrong with that one.

**Comment by rmcanany** [comment]
That one file is bombing out at line 875 of `UtilsCommon`.  
'dsiStream = cf.RootStorage.GetStream("DocumentSummaryInformation")'

It's probably messed up somehow.  Sent an email.

**Comment by rmcanany** [comment]
Is there a faster way of doing this?  I had populated the list with a large number of files, then accidentally hit Update again.  It took forever.
```
        For i = ListViewFiles.Items.Count - 1 To 0 Step -1
            If ListViewFiles.Items.Item(i).Group.Name <> "Sources" Then
                ListViewFiles.Items.Item(i).Remove()
            Else
                GroupTags.Add(CType(ListViewFiles.Items.Item(i).Tag, String))
            End If
        Next
```

I tried this, but got an exception adding an item to `tmpListViewFiles`.
```
        'Dim tmpListViewFiles As New ListView
        'Dim TagName As String

        'For i = ListViewFiles.Items.Count - 1 To 0 Step -1
        '    If ListViewFiles.Items.Item(i).Group.Name = "Sources" Then
        '        tmpListViewFiles.Items.Add(ListViewFiles.Items.Item(i))
        '    Else
        '        TagName = CType(ListViewFiles.Items.Item(i).Tag, String)
        '        If Not GroupTags.Contains(TagName) Then
        '            GroupTags.Add(TagName)
        '        End If
        '    End If
        'Next

        'ListViewFiles = tmpListViewFiles
```

**Comment by farfilli** [comment]
What If we create two separate list for files and sources? In this way we could just clear the listviewitems collection and recreate the source items from the list

**Comment by rmcanany** [comment]
Brilliant!

**Comment by rmcanany** [comment]
I can give it a go.  It's gotta be getting late over there.

**Comment by rmcanany** [comment]
Quick FYI, I switch to working on Doc Status and haven't touched the Listview.  If you want to work on it, go for it.  You shouldn't hit any merge conflicts.  If not, don't give it a second thought.  It's next on my list.

**Comment by rmcanany** [comment]
Agreed.  Added a new issue: https://github.com/rmcanany/SolidEdgeHousekeeper/issues/158

**Comment by farfilli** [comment]
@rmcanany can you check why the LVItem.Bounds return a wrong Bottom value?
I did a test with the ListViewCollapsible in another project, and it works correctly; therefore, something else should be in the initialization of the form.

The line of code is this: https://github.com/rmcanany/SolidEdgeHousekeeper/blob/a06ecad8e68df4a3888400c2278a733809b37ddb/My%20Project/UtilsFileList.vb#L1176

**Comment by rmcanany** [comment]
You're trying to determine visibility on the form, is that right?  I don't really see how that would help, but I trust you.

Anyway, I noticed that the rectangle was always 0,0,0,0 when the LV was first cleared, then repopulated.  That probably makes sense since the LVItem is added to ListViewFiles only after all the processing.

If I added a new property column, then clicked Close, I saw the values changing.  I inspected the bounds like so (with a break point, not shown, below these lines).
```
        'If LVItem.Bounds.IntersectsWith(FMain.ListViewFiles.ClientRectangle) Then
        Dim x = LVItem.Bounds
        Dim y = LVItem.GetBounds(ItemBoundsPortion.Entire)
        Dim z = FMain.ListViewFiles.ClientRectangle
```

You had previously mentioned only populating properties only for visible columns.  What about changing this:
```
If PropColumn.Name <> "Name" And PropColumn.Name <> "Path" Then
```

to this?
```
If PropColumn.Name <> "Name" And PropColumn.Name <> "Path" And PropColumn.Visible Then
```

**Comment by farfilli** [comment]
The point is to update ListviewItem.Subitems only on visible items.
In my tests the bounds.bottom property was always like -210000000 that obviously never intersect the listview bounds

**Comment by rmcanany** [comment]
I didn't see numbers like that.  I did get an exception if I tried to add a new property column with a section collapsed.

Wouldn't checking the column visibility do what you want, or am I missing something?  Also, what would happen if the user scrolled the list down after the update?  Would those properties not be present?

**Comment by farfilli** [comment]
I will take again a look tomorrow, my plan is to update only visible items when they appear.

**Comment by rmcanany** [comment]
Got it.  Somewhere I saw LV virtual mode mentioned in that context.

**Comment by rmcanany** [comment]
Took a stab at clearing ListViewFiles and repopulating just the Source items.  It's a little clunky.  Any streamlining ideas are welcome.

**Comment by farfilli** [comment]
I think we can close this as completed. 

With the latest code optimization the population of properties its quite fast and if an user has to work with thousands of file he can always disable the properties columns to speed up. 

Perhaps a warning when more than x thousands of file are found could pop-up?

**Comment by rmcanany** [comment]
I did add a warning a few weeks ago in `UtilsFileList.vb` in method `PopulateListView()`.  Is it not showing up for you?  It may be in the wrong spot.

```
    Public Sub PopulateListView(FoundFiles As IReadOnlyCollection(Of String))

        Dim UC As New UtilsCommon
        Dim tf As Boolean
        Dim NumProcessed As Integer = 0
        Dim PopulatePropertyColumns As Boolean = True

        ' ###### User input for displaying file properties for a large number of files.
        If (FMain.ListOfColumns.Count > 2) And (FoundFiles.Count > 1000) Then
            Dim s As String = String.Format("Getting file properties on {0} files can take some time.  ", FoundFiles.Count)
            s = String.Format("{0}Do you want to have them displayed anyway?", s)
            Dim Result As MsgBoxResult = MsgBox(s, vbYesNo)
            If Result = MsgBoxResult.No Then
                PopulatePropertyColumns = False
            End If
        End If

```

**Comment by farfilli** [comment]
Sorry Robert, I didn't know about it. Will check it tomorrow but seems ok

**Comment by rmcanany** [comment]
Closing as complete.  It is a great addition.  Thank you!

**Comment by rmcanany** [comment]
Closing as complete for real this time.
